### PR TITLE
DOCUMENTATION: v2.10.3 Abstraction Refinement

### PR DESCRIPTION
**Pull request by glhays**
Closes: #301
Closes: #173

### CODE REVIEW

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ For instance, if a storage broker is offering `InsertStudentAsync(Student studen
 ```csharp
 public async ValueTask<Student> AddStudentAsync(Student student)
 {
-	ValidateStudentOnAdd(student);
+	ValidateStudentOnAddAsync(student);
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:18]:
```suggestion
	await ValidateStudentOnAddAsync(student);
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 2]
```diff
@@ -53,7 +53,7 @@ But they offer a validation and exception handling (and mapping) wrapper around
 public ValueTask<Student> AddStudentAsync(Student student) =>
 TryCatch(async () =>
 {
-	ValidateStudentOnAdd(student);
+	ValidateStudentOnAddAsync(student);
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:56]:
```suggestion
	await ValidateStudentOnAddAsync(student);
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 3]
```diff
@@ -407,12 +407,14 @@ private void Validate(params (dynamic Rule, string Parameter)[] validations)
...
 {
+	......
+
 	Validate(
-		(Rule: IsInvalid(student.Id), Parameter: nameof(Student.Id)),
-		(Rule: IsInvalid(student.Name), Parameter: nameof(Student.Name)),
-		(Rule: IsInvalid(student.Grade), Parameter: nameof(Student.Grade))
+		(Rule: IsInvalidAsync(student.Id), Parameter: nameof(Student.Id)),
+		(Rule: IsInvalidAsync(student.Name), Parameter: nameof(Student.Name)),
+		(Rule: IsInvalidAsync(student.Grade), Parameter: nameof(Student.Grade))
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:417]:
```suggestion
		(Rule: await IsInvalidAsync(student.Id), Parameter: nameof(Student.Id)),
		(Rule: await IsInvalidAsync(student.Name), Parameter: nameof(Student.Name)),
		(Rule: await IsInvalidAsync(student.Grade), Parameter: nameof(Student.Grade))
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 4]
```diff
@@ -449,18 +451,22 @@ In this case, we can't validate `Address` as a property on the `Student` level a
...
 
+	......
+
 	Validate(
-		(Rule: IsInvalid(student.Address.Street), Parameter: nameof(StudentAddress.Street)),
-		(Rule: IsInvalid(student.Address.City), Parameter: nameof(StudentAddress.City)),
-		(Rule: IsInvalid(student.Address.ZipCode), Parameter: nameof(StudentAddress.ZipCode))
+		(Rule: IsInvalidAsync(student.Address.Street), Parameter: nameof(StudentAddress.Street)),
+		(Rule: IsInvalidAsync(student.Address.City), Parameter: nameof(StudentAddress.City)),
+		(Rule: IsInvalidAsync(student.Address.ZipCode), Parameter: nameof(StudentAddress.ZipCode))
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:469]:
```suggestion
		(Rule: await IsInvalidAsync(student.Id), Parameter: nameof(Student.Id)),
		(Rule: await IsInvalidAsync(student.Name), Parameter: nameof(Student.Name)),
		(Rule: await IsInvalidAsync(student.Address), Parameter: nameof(Student.Address))
	);

	......

	Validate(
		(Rule: await IsInvalidAsync(student.Address.Street), Parameter: nameof(StudentAddress.Street)),
		(Rule: await IsInvalidAsync(student.Address.City), Parameter: nameof(StudentAddress.City)),
		(Rule: await IsInvalidAsync(student.Address.ZipCode), Parameter: nameof(StudentAddress.ZipCode))
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 5]
```diff
@@ -621,15 +627,17 @@ Here's how an Id validation would look like:
 
 ```csharp
 
-private void ValidateStudentOnAdd(Student student)
+private async ValueTask ValidateStudentOnAddAsync(Student student)
 {
-    Validate((Rule: IsInvalid(student.Id), Parameter: nameof(Student.Id)));
+	......
+
+    Validate((Rule: IsInvalidAsync(student.Id), Parameter: nameof(Student.Id)));
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:634]:
```suggestion
    Validate((Rule: await IsInvalidAsync(student.Id), Parameter: nameof(Student.Id)));
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 6]
```diff
@@ -690,11 +698,11 @@ private async ValueTask<Student> TryCatch(ReturningStudentFunction returningStud
 	}
 	catch (InvalidStudentException invalidStudentInputException)
 	{
-		throw CreateAndLogValidationException(invalidStudentInputException);
+		throw await CreateAndLogValidationExceptionAsync(invalidStudentInputException);
 	}
 }
 
-private StudentValidationException CreateAndLogValidationException(Xeption exception)
+private async ValueTask<StudentValidationException> CreateAndLogValidationExceptionAsync(Xeption exception)
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:705]: missing await line 711:
`    await this.loggingBroker.LogErrorAsync(studentValidationException);`

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 7]
```diff
@@ -779,44 +809,205 @@ private async Task ShouldThrowValidationExceptionOnAddIfUpdatedByNotSameAsCreate
...
+};
+
+private async ValueTask<(bool IsNotRecent, DateTimeOffset StartDate, DateTimeOffset EndDate)>
+	IsDateNotRecentAsync(DateTimeOffset date)
+{
+	int pastSeconds = 60;
+	int futureSeconds = 0;
+
+	DateTimeOffset currentDateTime =
+		await this.dateTimeBroker.GetCurrentDateTimeOffsetAsync();
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:998]:
```suggestion
	DateTimeOffset currentDateTime =
		await this.dateTimeBroker.GetCurrentDateTimeOffsetAsync();

    if (currentDateTime == default)
    {
        return (false, default, default);
    }
```

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 8]
```diff
@@ -1126,16 +1331,18 @@ private async ValueTask<Student> TryCatch(ReturningStudentFunction returningStud
...
 
-        throw CreateAndLogDependencyValidationException(alreadyExistsStudentException);
+        throw await CreateAndLogDependencyValidationExceptionAsync(alreadyExistsStudentException);
     }
 }
 
 ...
 
-private StudentDependencyValidationException CreateAndLogDependencyValidationException(Xeption exception)
+private async ValueTask<StudentDependencyValidationException> CreateAndLogDependencyValidationException(
```
Comment (cjdutoit) [inline 2. Services/2.1 Foundations/2.1 Foundations.md:1344]:
```suggestion
private async ValueTask<StudentDependencyValidationException> CreateAndLogDependencyValidationExceptionAsync(
```

File: 2. Services/2.3 Orchestrations/2.3 Orchestrations.md [hunk 9]
```diff
@@ -720,23 +720,24 @@ public partial class StudentOrchestrationService
...
         }
         catch (LibraryCardDependencyValidationException libraryCardDependencyValidationException)
         {
-            throw CreateAndLogDependencyValidationException(libraryCardDependencyValidationException);
+            throw await CreateAndLogDependencyValidationExceptionAsync(libraryCardDependencyValidationException);
         }
     }
 
-    private StudentOrchestrationDependencyValidationException CreateAndLogDependencyValidationException(Xeption exception)
+    private async ValueTask<StudentOrchestrationDependencyValidationException>
```
Comment (cjdutoit) [inline 2. Services/2.3 Orchestrations/2.3 Orchestrations.md:739]: need the await on line 747:
`        await this.loggingBroker.LogErrorAsync(studentOrchestrationDependencyValidationException);`

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 1 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 2 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 3 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 4 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md [hunk 10]
```diff
@@ -595,7 +601,7 @@ public class StudentValidationException : Xeption
 The string messaging for the outer-validation above will be passed when the exception is initialized from the service class as shown below.
 
 ```csharp
-private StudentValidationException CreateAndLogValidationException(Xeption exception)
+private async ValueTask<StudentValidationException> CreateAndLogValidationExceptionAsync(Xeption exception)
```
[the remark this answers is not in the document]
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md:604]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 5 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 6 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md:705]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 7 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md:998]: Fixed

File: 2. Services/2.1 Foundations/2.1 Foundations.md (hunk 8 above)
Comment (glhays) [reply 2. Services/2.1 Foundations/2.1 Foundations.md]: Fixed

File: 2. Services/2.3 Orchestrations/2.3 Orchestrations.md (hunk 9 above)
Comment (glhays) [reply 2. Services/2.3 Orchestrations/2.3 Orchestrations.md:739]: Fixed

**Review by hassanhabib — APPROVED** [review]

**Pull request merged**